### PR TITLE
#173216044 add property endpoint

### PR DESCRIPTION
**Pull request by tobieeniafe**
#### What does this PR do?
Add property andpoint for dashboard

#### Description of Task to be completed?
Create endpoints to add, view, update & delete properties

#### How should this be manually tested?
`POST /api/v1/property/add`
`PUT /api/v1/property/update`
`DELETE /api/v1/property/delete/:id`
`GET /api/v1/property/all`
`GET /api/v1/property/:id`

`{
  "name": "5 bedroom terrace duplex",
  "location": "lagos",
  "price": 40000000,
  "units": 3,
  "houseType": "5 bedroom terrace duplex",
  "bedrooms": 5,
  "toilets": 7,
  "description": "Newly built 5 bedroom terrace duplex",
  "floorPlans": "http://linktoplan.ng",
  "mapLocation": {"longitude": "1.23456", "latitude": "2.34567"},
  "neighborhood": ["Lekki phase 1"],
  "mainImage": "https://picsum.photos/200",
  "gallery": ["https://picsum.photos/200", "https://picsum.photos/200", "https://picsum.photos/200"]
}`

#### Any background context you want to provide?
All requests should be sent with token as header `Authorization`

#### What are the relevant pivotal tracker stories?
#173216044

### CODE REVIEW

File: server/controllers/property.controllers.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  deleteProperty,
+  getPropertyByRefNo,
+  getOwnedProperties,
+} from '../services/property.service';
+
+const PropertyController = {
+  add(req, res, next) {
+    const property = req.locals;
+    const { user } = req;
+    property.owner = user._id;
```
Comment (andela-hpopoola) [inline server/controllers/property.controllers.js:13]: Can we try to merge this with the method invocation below? It is good practises to avoid object mutation.

File: server/controllers/property.controllers.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    addProperty(property)
```
Comment (andela-hpopoola) [inline server/controllers/property.controllers.js:14]:
```suggestion
    addProperty({...property, owner: user._id })
```

File: server/schemas/property.schema.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+export const propertySchema = Joi.object({
+  refNo: Joi.string().label('Reference number').required(),
+  category: Joi.string().label('Product category').required(),
+  type: Joi.string().label('Product type').required(),
+  price: Joi.number().label('Product price').required(),
+  area: Joi.string().label('Product area').required(),
+  state: Joi.string().label('Product state').required(),
+});
+
+export const propertyUpdateSchema = Joi.object({
```
Comment (andela-hpopoola) [inline server/schemas/property.schema.js:12]: Not sure why you are verifying only the email address for an updated property. I think we need to verify all the input submitted by the admin to be sure that they are valid.

File: server/controllers/property.controllers.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+      .catch((error) => next(error));
+  },
+  delete(req, res, next) {
+    const { refNo } = req.params;
+    const { user } = req;
+    const owner = user._id;
+
+    getPropertyByRefNo(refNo)
+      .then((property) => {
+        // eslint-disable-next-line eqeqeq
```
Comment (andela-hpopoola) [inline server/controllers/property.controllers.js:72]: Even though the code below will likely be removed, I see no reason to disable eslint here. If the code below doesn't give you the expected result in this case, then it means you are doing something wrong (for the values not be equal) and would need to check your code again.

File: server/models/user.model.js [hunk 5]
```diff
@@ -54,6 +54,13 @@ const UserSchema = new mongoose.Schema(
       select: false,
     },
     phone: String,
+    role: {
+      type: Number,
+      default: 1,
+    },
+    assignedProperties: {
+      type: [],
```
Comment (andela-hpopoola) [inline server/models/user.model.js:62]: This should be a list of ObjectIds. Something like this (You might need to confirm from mongoose documentation):
```suggestion
      type: [{ type : ObjectId, ref: 'Property' }],
```

File: server/controllers/property.controllers.js (hunk 4 above)
Comment (tobieeniafe) [reply server/controllers/property.controllers.js:72]: the values are equal, but the data types differ

Comment (andela-hpopoola) [reply server/controllers/property.controllers.js:72]: That is exactly my point sir, the values and the data type should be the same. If they are not the same, you are doing something wrong. From the code, it seems one is an ObjectId and the other is a Number. Both of them should be ObjectId in this case.

File: server/controllers/property.controllers.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+} from '../services/property.service';
+import httpStatus from '../helpers/httpStatus';
+
+const PropertyController = {
+  add(req, res, next) {
+    const property = req.locals;
+    const { user } = req;
+    addProperty({ ...property, addedBy: user._id, updatedBy: user._id })
+      .then(() => {
+        res.status(httpStatus.CREATED).json({ success: true, message: 'Property added' });
```
Comment (andela-hpopoola) [inline server/controllers/property.controllers.js:16]: Let's return the newly added property to the response.

File: server/helpers/middleware.js [hunk 7]
```diff
@@ -43,6 +44,18 @@ export const authenticate = async (req, res, next) => {
       error,
     });
   }
+  return null;
+};
 
+export const isAdmin = async (req, res, next) => {
+  const { user } = req;
+  if (user.role === userRole.ADMIN) {
```
Comment (andela-hpopoola) [inline server/helpers/middleware.js:52]: Lets verify if user is valid here, so that we can get an error if used without authenticating (i.e. if user is null)
```suggestion
  if (user && user.role === userRole.ADMIN) {
```

File: server/helpers/middleware.js [hunk 8]
```diff
@@ -43,6 +44,18 @@ export const authenticate = async (req, res, next) => {
...
+    next();
+  } else {
+    return res.status(httpStatus.FORBIDDEN).json({
+      success: false,
+      message: 'Action can only be performed by an Admin',
```
Comment (andela-hpopoola) [inline server/helpers/middleware.js:57]:
```suggestion
      message: 'You are not permitted to perform this action',
```

File: server/models/property.model.js [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    },
+    toilets: {
+      type: Number,
+      required: true,
+    },
+    description: {
+      type: String,
+      required: true,
+    },
+    floorPlans: {
```
Comment (andela-hpopoola) [inline server/models/property.model.js:100]: I think this will be a link to an uploaded floor plan document from the mockup. Let's make it a string for the main time.

File: server/models/property.model.js [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+      type: Boolean,
+      required: true,
```
Comment (andela-hpopoola) [inline server/models/property.model.js:102]: I think this is better as optional

File: server/models/property.model.js [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    },
+    mapLocation: {
+      type: String,
+      required: true,
+    },
+    neighborhood: {
```
Comment (andela-hpopoola) [inline server/models/property.model.js:108]: From the Mockup, this should be an array of strings.

File: server/models/property.model.js [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+      type: String,
+      required: true,
+    },
+    addedBy: {
+      type: String,
```
Comment (andela-hpopoola) [inline server/models/property.model.js:113]: This should be an ObjectId so that it can have the same data type as the user id.

File: server/models/property.model.js [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+      required: true,
+    },
+    updatedBy: {
+      type: String,
```
Comment (andela-hpopoola) [inline server/models/property.model.js:117]: Same comment as above.

File: server/models/user.model.js [hunk 14]
```diff
@@ -47,14 +47,38 @@ const UserSchema = new mongoose.Schema(
...
     },
     phone: String,
-    activated: { type: Boolean, default: false },
+    role: {
+      type: Number,
+      default: 1,
+    },
+    assignedProperties: [
+      {
+        propertyId: String,
```
Comment (andela-hpopoola) [inline server/models/user.model.js:67]: This should be an ObjectId imho

File: server/models/user.model.js [hunk 15]
```diff
@@ -47,14 +47,38 @@ const UserSchema = new mongoose.Schema(
...
+        assignedBy: String,
```
Comment (andela-hpopoola) [inline server/models/user.model.js:68]: Also this, ObjectId

File: server/models/user.model.js [hunk 16]
```diff
@@ -47,14 +47,38 @@ const UserSchema = new mongoose.Schema(
...
+        assignedDate: Date,
+        approved: {
+          type: Boolean,
+          default: false,
+        },
+        approvedBy: String,
```
Comment (andela-hpopoola) [inline server/models/user.model.js:74]: This too 😬

File: server/models/property.model.js [hunk 17]
```diff
@@ -98,23 +100,22 @@ const PropertySchema = new mongoose.Schema(
...
     },
     mapLocation: {
       type: String,
       required: true,
     },
     neighborhood: {
-      type: String,
+      type: [String],
       required: true,
```
Comment (andela-hpopoola) [inline server/models/property.model.js:109]: Please let's make this optional, please?

File: server/models/property.model.js [hunk 18]
```diff
@@ -98,23 +100,22 @@ const PropertySchema = new mongoose.Schema(
...
     },
     floorPlans: {
-      type: Boolean,
-      required: true,
+      type: String,
     },
     mapLocation: {
       type: String,
       required: true,
```
Comment (andela-hpopoola) [inline server/models/property.model.js]: Lets make this optional too?

File: server/models/property.model.js [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+ *          - location
+ *          - price
+ *          - units
+ *          - houseType
+ *          - bedrooms
+ *          - toilets
+ *          - description
+ *          - floorPlans
+ *          - mapLocation
+ *          - neighborhood
```
Comment (andela-hpopoola) [inline server/models/property.model.js:20]: We need to add the `property image` and `gallery` to the property schema. I mentioned you(your highrachy mail) on the spreadsheet document when I noticed that it was missing.

File: server/schemas/property.schema.js [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+  name: Joi.string().label('Property name').required(),
+  location: Joi.string().label('Property location').required(),
+  price: Joi.number().label('Property price').required(),
+  units: Joi.number().label('Property units').required(),
+  houseType: Joi.string().label('Property type').required(),
+  bedrooms: Joi.number().label('Bedroom number').required(),
+  toilets: Joi.number().label('Toilet number').required(),
+  description: Joi.string().label('Property description').required(),
+  floorPlans: Joi.boolean().label('Property floor plans').required(),
+  mapLocation: Joi.string().label('Map location').required(),
```
Comment (andela-hpopoola) [inline server/schemas/property.schema.js:13]: This should test for an optional object

File: server/schemas/property.schema.js [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+  neighborhood: Joi.string().label('Map location').required(),
+});
+
+export const updatePropertySchema = Joi.object({
```
Comment (andela-hpopoola) [inline server/schemas/property.schema.js:17]: Instead of repeating the same values here as above, can you assign the same values into an object and spread it in the two objects to ensure that updates are made in one place.

File: server/schemas/property.schema.js [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+  location: Joi.string().label('Property location').required(),
+  price: Joi.number().label('Property price').required(),
+  units: Joi.number().label('Property units').required(),
+  houseType: Joi.string().label('Property type').required(),
+  bedrooms: Joi.number().label('Bedroom number').required(),
+  toilets: Joi.number().label('Toilet number').required(),
+  description: Joi.string().label('Property description').required(),
+  floorPlans: Joi.boolean().label('Property floor plans').required(),
+  mapLocation: Joi.string().label('Map location').required(),
+  neighborhood: Joi.string().label('Map location').required(),
```
Comment (andela-hpopoola) [inline server/schemas/property.schema.js:14]: This should test for an optional array

File: server/schemas/property.schema.js [hunk 23]
```diff
@@ -9,22 +14,16 @@ export const addPropertySchema = Joi.object({
...
   toilets: Joi.number().label('Toilet number').required(),
   description: Joi.string().label('Property description').required(),
-  floorPlans: Joi.boolean().label('Property floor plans').required(),
-  mapLocation: Joi.string().label('Map location').required(),
-  neighborhood: Joi.string().label('Map location').required(),
-});
+  floorPlans: Joi.string().label('Property floor plans'),
+  mapLocation,
+  neighborhood: Joi.array().label('Property neighborhood'),
+  mainImage: Joi.string().label('Property neighborhood').required(),
```
Comment (andela-hpopoola) [inline server/schemas/property.schema.js:20]: Let's make the main image and gallery optional, please.

File: server/models/property.model.js [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+    description: {
+      type: String,
+      required: true,
+    },
+    floorPlans: {
+      type: String,
+    },
+    mapLocation: {
+      longitude: {
+        type: String,
```
Comment (andela-hpopoola) [inline server/models/property.model.js:107]: Can we make this a number instead of a string so that it tallies with the example you gave in the API docs?

File: server/controllers/property.controllers.js [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+} from '../services/property.service';
+import httpStatus from '../helpers/httpStatus';
+
+const PropertyController = {
+  add(req, res, next) {
+    const property = req.locals;
+    const { user } = req;
+    addProperty({ ...property, addedBy: user._id, updatedBy: user._id })
+      .then(() => {
+        res.status(httpStatus.CREATED).json({ success: true, message: 'Property added', property });
```
Comment (andela-hpopoola) [inline server/controllers/property.controllers.js:16]: Can you return the property returned by the database instead of the one submitted by the user. I think it is preferable for the returned property to contain an id (even if we want to omit other things).

File: server/schemas/property.schema.js [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+  floorPlans: Joi.string().label('Property floor plans').optional(),
+  mapLocation,
+  neighborhood: Joi.array().label('Property neighborhood').optional(),
+  mainImage: Joi.string().label('Property neighborhood').optional(),
+  gallery: Joi.array().label('Property neighborhood').optional(),
+};
+
+export const addPropertySchema = Joi.object({ ...product });
+
+export const updatePropertySchema = Joi.object({
```
Comment (andela-hpopoola) [inline server/schemas/property.schema.js:26]: I think I am wrong here. We should be able to update a property without stating all the values in the property. We should be able to update only one section (maybe the units with the id alone without stating other properties).

Lets readd the properties and make them optional instead of required so that the schema validation will allow us to add a subset of the property instead of adding more.

File: server/schemas/property.schema.js [hunk 27]
```diff
@@ -25,5 +25,17 @@ export const addPropertySchema = Joi.object({ ...product });
 
 export const updatePropertySchema = Joi.object({
   id: Joi.string().label('Property id').required(),
-  ...product,
+  units: Joi.number().label('Property units').required(),
```
Comment (andela-hpopoola) [inline server/schemas/property.schema.js:28]: Not sure why this is still `required`. I thought we wanted to make them optional?

File: test/services/property.service.test.js [hunk 28]
```diff
@@ -0,0 +1,63 @@
+import mongoose from 'mongoose';
+import { expect, useDatabase } from '../config';
+import {
+  getPropertyById,
+  addProperty,
+  //   updateProperty,
```
Comment (andela-hpopoola) [inline test/services/property.service.test.js:6]: I am not sure why you commented this out, we need to write tests to check for property updates, delete and getAllProperties.

File: test/factories/property.factory.js [hunk 29]
```diff
@@ -0,0 +1,19 @@
+import { Factory } from 'rosie';
+
+export default new Factory()
```
Comment (andela-hpopoola) [inline test/factories/property.factory.js:3]: Can you update the factory so that it contains the most updated version of property schema. I noticed that mainImage and gallery fields are missing.

File: test/controllers/property.controller.test.js [hunk 30]
```diff
@@ -0,0 +1,296 @@
...
+
+describe('Update Property', () => {
+  const id = mongoose.Types.ObjectId();
+  const property = PropertyFactory.build({ _id: id, addedBy: _id, updatedBy: _id });
+
+  beforeEach(async () => {
+    await addProperty(property);
+  });
+
+  const newProperty = {
```
Comment (andela-hpopoola) [inline test/controllers/property.controller.test.js:65]: This might look like I am asking for too much, but to ensure that our test works in the best way for us, we also need to test for all that we can update all the fields one by one with our test (not a part of it). Updating just a part of it works to test that the update controller works, however, we also need to test that each field will not cause an error / bug and would properly be updated when they are done by the user.

Can you add the other fields in this object and test that all of them are properly updated?

File: test/services/property.service.test.js [hunk 31]
```diff
@@ -0,0 +1,157 @@
+import mongoose from 'mongoose';
+import { expect, sinon, useDatabase } from '../config';
+import {
+  getPropertyById,
+  addProperty,
+  updateProperty,
+  deleteProperty,
+  // getAllProperties,
```
Comment (andela-hpopoola) [inline test/services/property.service.test.js:8]: Not sure why you commented this out? Can we add a basic test for it, please?

**Comment by tobieeniafe** [comment]
> Thank you for working on this @tobieeniafe. We are also through with this, thank you for your patience and endurance (I know adding tests can sometimes but frustrating).
> 
> I added a few more comments on updating a few tests. Also, I didn't see any implementation for this as stated in one of my comments above.
> 
> > getAllProperties endpoint needs to be properly fleshed out. The current implementation is not usable and it will be better if we can split it to the following:
> > a. getAllUserProperties - This should get all the properties owned by a user. Each property should contain the admin details.
> > b. getAllAdminAddedProperties - Get all the properties added by an admin. Each property should have the user information that has also been assigned to the property.
> > c. getAllProperties: Get all properties in the database with the assigned user and admin details.
> 
> I am happy if you want to implement them on another Pull Request. Just inform me and I will be glad to create a new story for it. Thank you and Nice Progress.
> 
> 

Let's add this to another PR please

File: server/schemas/property.schema.js [hunk 32]
```diff
@@ -0,0 +1,41 @@
...
+  price: Joi.number().label('Property price').required(),
+  units: Joi.number().label('Property units').required(),
+  houseType: Joi.string().label('Property type').required(),
+  bedrooms: Joi.number().label('Bedroom number').required(),
+  toilets: Joi.number().label('Toilet number').required(),
+  description: Joi.string().label('Property description').required(),
+  floorPlans: Joi.string().label('Property floor plans').optional(),
+  mapLocation,
+  neighborhood: Joi.array().label('Property neighborhood').optional(),
+  mainImage: Joi.string().label('Property mainImage').optional(),
```
Comment (andela-hpopoola) [inline server/schemas/property.schema.js:20]:
```suggestion
  mainImage: Joi.string().label('Property main image').optional(),
```

File: server/schemas/property.schema.js [hunk 33]
```diff
@@ -0,0 +1,41 @@
...
+  location: Joi.string().label('Property location').optional(),
+  price: Joi.number().label('Property price').optional(),
+  houseType: Joi.string().label('Property type').optional(),
+  bedrooms: Joi.number().label('Bedroom number').optional(),
+  toilets: Joi.number().label('Toilet number').optional(),
+  description: Joi.string().label('Property description').optional(),
+  floorPlans: Joi.string().label('Property floor plans').optional(),
+  mapLocation,
+  neighborhood: Joi.array().label('Property neighborhood').optional(),
+  mainImage: Joi.string().label('Property mainImage').optional(),
```
Comment (andela-hpopoola) [inline server/schemas/property.schema.js:39]:
```suggestion
  mainImage: Joi.string().label('Property main image').optional(),
```

File: test/controllers/property.controller.test.js [hunk 34]
```diff
@@ -51,28 +90,235 @@ describe('Add Property Route', () => {
...
+        const property = PropertyFactory.build({ mainImage: '' });
+        request()
+          .post('/api/v1/property/add')
+          .set('authorization', adminToken)
+          .send(property)
+          .end((err, res) => {
+            expect(res).to.have.status(412);
+            expect(res.body.success).to.be.eql(false);
+            expect(res.body.message).to.be.eql('Validation Error');
+            expect(res.body.error).to.be.eql('"Property mainImage" is not allowed to be empty');
```
Comment (andela-hpopoola) [inline test/controllers/property.controller.test.js:248]: Based on the above nits:
```suggestion
            expect(res.body.error).to.be.eql('"Property main image" is not allowed to be empty');
```

File: test/controllers/property.controller.test.js [hunk 35]
```diff
@@ -51,28 +90,235 @@ describe('Add Property Route', () => {
...
+            done();
+          });
+      });
+    });
+    context('when gallery is empty', () => {
```
Comment (andela-hpopoola) [inline test/controllers/property.controller.test.js:253]: The gallery is supposed to be an array, so we need to add other tests to test that we can add an empty array. The same should apply for other optional fields.

File: test/controllers/property.controller.test.js [hunk 36]
```diff
@@ -230,12 +739,29 @@ describe('Get all properties', () => {
...
     });
 
+    context('when user token is used', () => {
+      beforeEach(async () => {
+        await User.findByIdAndDelete(_id);
+      });
+      it('returns token error', (done) => {
+        request()
+          .get('/api/v1/property/all')
+          .set('authorization', adminToken)
```
Comment (andela-hpopoola) [inline test/controllers/property.controller.test.js:749]: The context states that `when user token is used`, however, you used the adminToken here.

File: test/controllers/property.controller.test.js [hunk 37]
```diff
@@ -268,11 +794,28 @@ describe('Delete property', () => {
...
   });
 
+  context('when user token is used', () => {
+    beforeEach(async () => {
+      await User.findByIdAndDelete(_id);
+    });
+    it('returns token error', (done) => {
+      request()
+        .delete(`/api/v1/property/delete/${id}`)
+        .set('authorization', adminToken)
```
Comment (andela-hpopoola) [inline test/controllers/property.controller.test.js:804]: Same comment as above

File: test/services/property.service.test.js [hunk 38]
```diff
@@ -154,4 +154,17 @@ describe('Property Service', () => {
...
+  describe('#getAllProperties', () => {
+    const _id = mongoose.Types.ObjectId();
+
+    before(async () => {
+      await Property.create(PropertyFactory.build({ _id, addedBy: _id, updatedBy: _id }));
+    });
+
+    it('returns a valid property by Id', async () => {
+      const property = await getAllProperties();
+      expect(property).to.be.an('array');
```
Comment (andela-hpopoola) [inline test/services/property.service.test.js:167]: Though we are fixing this in another PR, however, testing that the response is an array is not a valid way to test that all the properties are returned. Here are my thought process:

1. Find a way to add 2 properties into the database for this test
2. Get all the properties and test that it returns all the properties
**(Optional tests but nice to have as they gives more confidence)**
3. Test that the returned properties are the same as the added ones
4. Add one more property
5. Verify that getAllProperties returns 3 properties

**Comment by andela-hpopoola** [comment]
Thank you for working on this. I found a few nits but they are nothing too serious. Feel free to merge this PR after you resolve them. Thank you and Nice work.
